### PR TITLE
Enable privileged containers for apiserver and controller

### LINKED ISSUE #588: Openstack cloud provider causing apiserver container to fail
## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): 1.8.4

**Environment**:
- **Kubernetes version** (use `kubectl version`): 1.8.4
- **Cloud provider or hardware configuration**: openstack icehouse
- **OS** (e.g. from /etc/os-release): CentOS 7.4
- **Kernel** (e.g. `uname -a`): Linux 3.10
- **Others**: Docker 17.09.0-ce

## What happened?
When adding ```--cloud-provider=openstack``` and ```--cloud-config=/etc/kubernetes/cloud.conf``` to ```/etc/kubernetes/manifests/kube-apiserver.yaml``` I get the following error on startup of the container:

```
[mount_linux.go:142] Mount failed: exit status 1
Mounting arguments: -t vfat -o ro /tmp/configdrive071298910
Output: mount: can't read '/etc/fstab': No such file or directory
Error mounting configdrive: mount failed
```

## What you expected to happen?
For the container to find the config-2 drive and stay alive

## How to reproduce it (as minimally and precisely as possible)?

```/etc/kubernetes/manifests/kube-apiserver.yaml``` and ```/etc/kubernetes/manifests/kube-controller-manager.yaml``` share the following:

```
spec:
  containers:
  - command:
    - --cloud-provider=openstack
    - --cloud-config=/etc/kubernetes/cloud.conf
    ...

volumeMounts:
- mountPath: /etc/kubernetes/cloud.conf
  name: cloud-config
  readOnly: true
...

volumes:
- hostPath:
    path: /etc/kubernetes/cloud.conf
    type: File
  name: cloud-config
...
```

```/etc/systemd/system/kubelet.service.d/10-kubeadm.conf```

```
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --cloud-provider=openstack --cloud-config=/etc/kubernetes/cloud.conf"
Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin"
Environment="KUBELET_DNS_ARGS=--cluster-dns=10.96.0.10 --cluster-domain=cluster.local"
Environment="KUBELET_AUTHZ_ARGS=--authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt"
Environment="KUBELET_CADVISOR_ARGS=--cadvisor-port=0"
Environment="KUBELET_CGROUP_ARGS=--cgroup-driver=cgroupfs"
Environment="KUBELET_CERTIFICATE_ARGS=--rotate-certificates=true --cert-dir=/var/lib/kubelet/pki"
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_AUTHZ_ARGS $KUBELET_CADVISOR_ARGS $KUBELET_CGROUP_ARGS $KUBELET_CERTIFICATE_ARGS $KUBELET_EXTRA_ARGS
```

```/etc/kubernetes/cloud.conf```

```
[Global]
auth-url=...
tenant-id=...
username=...
password=...
region=...
ca-file=...
```

## Anything else we need to know?

The openstack metadata url is not available as HTTP, so unless HTTPS support is added, I need the config drive to work as expected.

Referred here from: https://github.com/kubernetes/kubernetes/issues/47392

### LINKED ISSUE #47392: kubeadm controller-manager no /etc/fstab when use cloud-provider "openstack" "config-drive"
Dear All:
I'm not sure if I'm right. When i use `kubeadm init --config=xxxx` whith `openstack provider config-drive`.I found controller-manager logs 
```
E0613 02:17:16.809590       1 mount_linux.go:122] Mount failed: exit status 1
Mounting command: mount
Mounting arguments:  /tmp/configdrive336243194 vfat [ro]
Output: mount: can't read '/etc/fstab': No such file or directory
``` 

#### env
```
kubeadm: 1.6.3
host: ubuntu xenial
network: calico ipip
```

Someone can help me? Thanks a lot.

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

In OpenStack environment, when there is no metadata service, we
look at the config drive to figure out the metadata. Since we need
to run commands like blkid, we need to ensure that api server and
kube controller are running in the privileged mode.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #47392
Fixes https://github.com/kubernetes/kubeadm/issues/588

**Special notes for your reviewer**:

**Release note**:

```release-note
Fix issue when using OpenStack config drive for node metadata
```

### CODE REVIEW

**Comment by kad** [comment]
Two questions: 1. unit tests. 2. can it be done with allowing few needed caps instead of fully privileged ?

**Comment by dims** [comment]
@kad we need CAP_SYS_ADMIN as we are mounting/unmounting file systems (similar to http://ceph.com/planet/no-more-privileged-containers-for-ceph-osds/). Any suggestions on the best way to do this?

Thanks for the quick review!

**Comment by kad** [comment]
In case of mounts, probably no suggestions from me at the moment. If CAP_SYS_ADMIN is enough, maybe that would be the best way.

**Comment by dims** [comment]
@kad so leave the current implementation as-is?

**Comment by kad** [comment]
@dims test case still will be good. so if someone will do refactoring on that code later, we will not loose it.
@liggitt what do you think about security/privileged pod for api manager and controller manager for that usage scenario ?

**Comment by dims** [comment]
@kad Done!

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go
Comment (kad) [inline cmd/kubeadm/app/phases/controlplane/manifests_test.go]: v1.9.0 ?

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go
Comment (kad) [inline cmd/kubeadm/app/phases/controlplane/manifests_test.go]: what about negative test scenario: privileged should be only if cloud provider == "openstack", but if not, it should be non-privileged.

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests_test.go]: ack

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go
Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests_test.go]: ack

**Comment by kad** [comment]
lgtm, however it will be great if @liggit or somebody else from k8s security specialists comment on privileged mode.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (liggitt) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Is this the first cloud provider specific things kubeadm is doing? I would like to see the plan for this going forward… this is just moving compiled-in logic up a level, when what we really want is to enable external integration of cloud provider specific modifications.

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @liggitt there's at lease one precedent:

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/controlplane/manifests.go#L174

Comment (liggitt) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Interesting. The linked example seems like a bug… admission plugins don't depend solely on cloud provider… seems like specifying cloud provider aws with kube 1.8 would not work with that code

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: https://godoc.org/k8s.io/kubernetes/pkg/util/pointer

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Done!

**Comment by dims** [comment]
cc @kubernetes/sig-cluster-lifecycle-pr-reviews

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: might be worth making `openstack` a [constant](https://github.com/kubernetes/kubernetes/tree/master/cmd/kubeadm/app/constants) and also referencing different providers in the unit tests

**Review by jamiehannaford — COMMENTED** [review]
I'd also like if we could reduce the permission scope if possible. If it's only possible with privileged, fine with me

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @jamiehannaford i see `"aws"` and `"gce"` as well. i'll do it in another PR after this one merges

Comment (jamiehannaford) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: okay sgtm

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Why does the controller manager need to talk with the meta-data service?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: I'd rather this be more generic enablement, b/c this isn't openstack specific.  This is a specific os-deployment issue.

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Should i remove the if condition for openstack?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: when the cloud provider is looking for "which node am i" info, in this specific scenario there is no openstack metadata service and the code looks in the config drive, to be able to mount the config drive it needs privs. we have a `search-order` (https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/) but in this case there is simply no metadata service and the only choice is the config service where it fails because the kube controller manager does not have enough super powers

Comment (kargakis) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: It would be nice to put this down into a code comment.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (jamiehannaford) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Do AWS and GCE also use the config drive for node metadata?

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: I don't know for sure @jamiehannaford

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @kargakis ack, if this looks ok otherwise, will happily do that in a little bit

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @dims I'd prefer an opt-in knob to enable privileged (default false), b/c again, this is very specific and not necessarily the common case.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Once the cloud-controller code is excised from the controller manager, is there any reason that we would still need this?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (jbeda) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Agree with a specific opt-in knob.  I'd consider it acceptable to have documentation for that knob state that often times cloud providers need it to be set and name openstack specifically.

Comment (jbeda) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Agree with a specific opt-in knob.  I'd consider it acceptable to have documentation for that knob state that often times cloud providers need it to be set and name openstack specifically.

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @timothysc correct, when we excise code that calls out to openstack cloud provider from these 2 containers we can yank the privilege flag.

Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: When is the ETA on that excise?  1.10, 1.11?  "Hand-wavy" future?

Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Also, would https://github.com/kubernetes/kubernetes/pull/58080 work for this use case?

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @timothysc we con't currently expose `--cloud-provider` from kubeadm command line, so that won't help. i've updated the PR with a new field in `MasterConfiguration` and updated the unit test. hopefully that combination will work for everyone.

Comment (dims) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @timothysc For 1.10, i do not expect to move at least the openstack cloud provider out of tree. at best we will have an external openstack controller manager that will work for some scenarios with the in-tree one still being the primary one. 1.11 is a distinct possibility.

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: The CloudProvider strings don't really need to be shuffled around, it makes it look like you're changing it for some reason.

**Review by timothysc — CHANGES_REQUESTED** [review]
minor nit on extraneous shufflings, then lgtm

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (dims) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Fixed. thanks!

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by dims** [comment]
@timothysc apologies, looks like i left one test broken, fixed it now. will ping when everything goes green.

**Comment by bputt** [comment]
Will this be in a 1.9.x release?

**Comment by errordeveloper** [comment]
@bputt I think this will land in 1.10 and won't get cherry-picked into 1.9.

**Comment by dims** [comment]
@errordeveloper @bputt Ilya is correct, this will fall into the new feature category (not bug fix)

**Pull request merged**
